showkase-component-impressions/src/main/java/com/vinaygaba/showkase_component_impressions/ImpressionCollector.kt

```diff
@@ -0,0 +1,55 @@
+package com.vinaygaba.showkase_component_impressions
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.launch
+
+internal class ImpressionCollector<T>(
+    val coroutineScope: CoroutineScope,
+    val onVisibilityEvent: (ShowkaseVisibilityEvent<T>) -> Unit
+) {
+    private val mutableImpressionEvents = MutableStateFlow<ShowkaseVisibilityEvent<T>?>(null)
+
+    @OptIn(FlowPreview::class)
+    val impressionEvents: Flow<ShowkaseVisibilityEvent<T>> =
+        mutableImpressionEvents.filterNotNull().debounce(200)
+
+    internal fun onLayoutCoordinatesChanged(
+        key: T,
+        visibilityPercentage: VisibilityMetadata,
+    ) {
+        publishImpressionEvent(key, visibilityPercentage)
+    }
+
+    internal fun publishImpressionEvent(
+        key: T,
+        visibilityMetadata: VisibilityMetadata,
+    ) {
+        coroutineScope.launch {
```

Comment:
I think right now this scope is not keyed so this will not be properly cancelled

Reply:
Since I got this coroutine scope using `rememberCoroutineScope`, its tied to the scope of the composition and will be auto-cancelled as soon as this composable is not in composition.